File: leetcodeInterviewQuestions/Question1514/Question1514.py

class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start_node: int, end_node: int) -> float:
        graph = defaultdict(list)
        for (a, b), prob in zip(edges, succProb):
            graph[a].append((b, prob))
            graph[b].append((a, prob))
        max_heap = [(-1.0, start_node)]
        max_prob = [0.0] * n
        max_prob[start_node] = 1.0
        
        while max_heap:
            current_prob, node = heapq.heappop(max_heap)
            current_prob = -current_prob
            if node == end_node:
                return current_prob
            for neighbor, prob in graph[node]:
                new_prob = current_prob * prob
                if new_prob > max_prob[neighbor]:
                    max_prob[neighbor] = new_prob
                    heapq.heappush(max_heap, (-new_prob, neighbor))
        return 0.0

File: leetcodeInterviewQuestions/Question1514/Question1514.cpp
class Solution
{
public:
      double maxProbability(int n, vector<vector<int>> &edges, vector<double> &succProb, int start, int end)
      {

            vector<vector<pair<int, double>>> g(n);
            for (int i = 0; i < edges.size(); i++)
            {
                  int u = edges[i][0], v = edges[i][1];
                  g[u].push_back({v, succProb[i]});
                  g[v].push_back({u, succProb[i]});
            }
            priority_queue<pair<double, int>> pq;
            vector<double> dist(n, 0.0);
            pq.push({1.0, start});
            dist[start] = 1.0;
            double cnt = 1.0;
            while (!pq.empty())
            {
                  double prob = pq.top().first;
                  int node = pq.top().second;
                  pq.pop();
                  // cnt *= prob;
                  // if(node == end){
                  //     return cnt;
                  // }
                  for (auto &it : g[node])
                  {
                        int curr = it.first;
                        double p = it.second;
                        if (dist[curr] < dist[node] * p)
                        {
                              dist[curr] = dist[node] * p;
                              pq.push({dist[curr], curr});
                        }
                  }
            }
            return dist[end];
      }
};